src/provider.rs

```diff
@@ -174,6 +197,164 @@ impl TableProvider for SeafowlTable {
+                field.name().clone(),
+                vec![ScalarValue::Utf8(None); partition_count],
+            );
+            null_counts.insert(field.name().clone(), vec![None; partition_count]);
+        }
+
+        for (ind, partition) in partitions.iter().enumerate() {
+            for column in partition.columns.as_slice() {
+                let data_type = schema
+                    .field_with_name(column.name.deref())
+                    .unwrap()
+                    .data_type();
+
+                min_values.get_mut(column.name.as_ref()).unwrap()[ind] =
+                    Self::parse_bytes_value(&column.min_value, data_type)?;
+                max_values.get_mut(column.name.as_ref()).unwrap()[ind] =
+                    Self::parse_bytes_value(&column.max_value, data_type)?;
+                null_counts.get_mut(column.name.as_ref()).unwrap()[ind] =
+                    column.null_count;
+            }
+        }
+
+        Ok(Self {
+            partition_count,
+            schema,
+            partitions: Arc::new(partitions),
+            min_values,
+            max_values,
+            null_counts,
+        })
+    }
+
+    // Try to deserialize min/max statistics stored as raw bytes
+    fn parse_bytes_value(
+        bytes_value: &Arc<Option<Vec<u8>>>,
+        data_type: &DataType,
+    ) -> Result<ScalarValue> {
+        match bytes_value.as_ref() {
+            Some(bytes) => match String::from_utf8(bytes.clone()) {
+                Ok(string_val) => ScalarValue::try_from_string(string_val, data_type),
+                Err(error) => Err(DataFusionError::Internal(format!(
+                    "Failed to parse min/max value: {}",
+                    error
+                ))),
+            },
+            None => Ok(ScalarValue::Utf8(None)),
+        }
+    }
+
+    // Prune away partitions that are refuted by the provided filter expressions
+    async fn prune(&self, filters: &[Expr]) -> Vec<SeafowlPartition> {
+        let mut partitions = self.partitions.deref().clone();
+
+        if filters.is_empty() {
+            // No qualifiers means we need all partitions
+            return partitions;
+        }
+
+        let mut partition_mask = vec![false; self.partition_count];
+
+        for expr in filters {
+            match PruningPredicate::try_new(expr.clone(), self.schema.clone()) {
+                Ok(predicate) => match predicate.prune(self) {
+                    Ok(expr_mask) => {
+                        partition_mask = partition_mask
+                            .iter()
+                            .zip(expr_mask)
+                            .map(|(&a, b)| a || b)
+                            .collect()
+                    }
+                    Err(error) => {
+                        warn!(
+                            "Failed pruning the partitions for expr {:?}: {}",
+                            expr, error
+                        );
+                        return partitions;
```

Comment:
Could continue here, since the next expr might not fail?

Reply:
We can't, because this particular filter may require a partition that no other filter does, thus skipping the partitions for this filter may lead to incorrect results. For example, for three partitions and two filters:

```python
expr1: [1, 0, 1] # first filter refutes the 2nd partition and needs the 1st and the 3rd one
expr2: [1, 1, 0] # second filter needs only the first two partitions
```
if pruning fails for the first filter and we continue, the third partition will not be included in the final partition vector.

Reply:
Actually, as discussed offline, you're correct; since the filters list is a big `AND` we can and should continue here (and flip the original partition mask to all 1's while using `&&` to combine the masks).